다이내믹 테마, 모바일 버거 메뉴 추가

sungik-choi opened the pull request:
## 기능 추가

- 다이내믹 테마(utterances, codeblock 포함)
- 모바일 버거 메뉴 추가

sungik-choi committed:
Feat : 링크 내부의 버튼 제거 후 링크 스타일 버튼으로 변경

- 테마 분리할 수 있는 부분들 분리
- aria-label 영어로 변경 (나중에 패키지 만들어 배포할 때 고려)

sungik-choi committed:
Refactor : 페이지네이션 보류, 불필요한 코드 삭제

- 변수명 일부 변경

sungik-choi committed:
Feat : 다크/라이트 테마 사용을 위한 커스텀 훅 작성

sungik-choi committed:
Refactor : 기존 테마 lightTheme로 변경

sungik-choi committed:
Feat : useTheme 사용해서 테마 불러오기

- useTheme 쪽이 2번 실행되면서 마지막 실행에 theme에 undefined가 담기는 문제가 있음.
- 수정 필요

sungik-choi committed:
Fix : useEffect -> useLayoutEffect 변경

- useEffect 훅 사용시 페이지 리프레쉬 때 깜빡임 현상이 나타남
- useLayoutEffect 사용시 이러한 현상 사라짐. 차이점을 더 알아봐야겠다.

sungik-choi committed:
Feat : eslint-react-hooks 플러그인 설치

sungik-choi committed:
Feat : 다크모드 테마 생성 (스타일링 진행 안됨)

sungik-choi committed:
Feat : 마크다운 h3, pre 스타일 변경

sungik-choi committed:
Style : 불필요한 코드 제거 및 상수 상단으로 이동

sungik-choi committed:
Style : 불필요한 코드 제거

sungik-choi committed:
Feat : theme 상태 context api로 child에게 전달

- utternances 테마를 변경하는 과정에서, 동일한 theme 상태를 유지해야 하는 이슈가 발생했다.
- useTheme() 커스텀 훅을 통해 가져온 theme 변수를 ThemeContext.Provider에 담아서 comment.js 파일에서도 동일한 theme에 접근할 수 있도록 했음
- layout 컴포넌트를 사용하는 예제들 중 children에게 props로 넘겨주는 방법들을 알아봤으나 context 쪽이 가장 쉬운 거 같아 이 방법을 사용

sungik-choi committed:
Feat : utternances에 다이나믹 테마 적용

- useContext를 통해 theme 변수를 사용한다
- siteMetaData에 theme가 필요없어졌으므로 함께 삭제 (useTheme에서 선호하는 테마를 확인하고 자동으로 변경해서 적용함)
- useLayoutEffect 첫번째 줄 : 현재 로컬스토리지에 저장돼있는 테마와 비교해서 같지 않으면 리턴한다.
- 이유 : setState가 비동기적으로 실행되어 theme가 원치 않는 테마인 채로 밸류값으로 들어가 utternaces가 생성됨. 의존성 배열에 theme가 들어가있으므로, theme가 제대로 변경됐을 때(변경됐다면 로컬스토리지에 그 값이 저장되도록 설정했다) utterances 생성 속성에도 동일한 테마가 들어가도록 설정
- utternaces 엘리먼트를 문서에서 찾고, 없다면 생성하고, 있다면 postMessage 메서드를 통해 iframe 내부로 theme 변경 메세지를 전달한다. theme 를 변경하는 건 layout 컴포넌트의 테마 토글러 버튼밖에 없으므로, 테마변경 버튼이 클릭되고 theme가 변경되면 테마 변경 메세지도 같이 전달된다.
- iframe이라그런지 동기적으로 자연스럽게 테마 변경이 되진 않지만 이정도도 만족스러움

sungik-choi committed:
Docs : 스타일드 컴포넌트로 디자인 시스템 따라하기 - 1 포스팅 추가

sungik-choi committed:
Feat : 상수 파일 분리

sungik-choi committed:
Feat : 다크 모드 스타일 적용

sungik-choi committed:
Feat : 스타일 다듬기

sungik-choi committed:
Feat : github-dark-theme 설치

sungik-choi committed:
Docs : 오탈자 수정

sungik-choi committed:
Style : 스타일 변수 수정

sungik-choi committed:
Feat : github-vscode-remark 플러그인 다이내믹 테마 적용

- useTheme 커스텀훅에 dataset을 html 태그에 추가하는 코드를 추가
- vscode-remark 플러그인에서 css에 접근해 테마를 변경할 수 있는 옵션 코드 추가

sungik-choi committed:
Feat : 테마 기본값 변경

sungik-choi committed:
Feat : focus에 따른 css 스타일 수정/변경, themeToggleButton 컴포넌트 추가

sungik-choi committed:
Docs : 예제 문서 추가

sungik-choi committed:
Fix : 잘못된 스타일 수정

sungik-choi committed:
Feat : 카테고리 선택 시 가로 스크롤 가운데로 오는 기능 추가

sungik-choi committed:
Feat : 카테고리 포커스 스타일 추가

- 아웃라인이 overflow:hidden 때문에 깨져 보이는 문제가 있어서 수정

sungik-choi committed:
Feat : navBar 테마 토글 버튼 추가

- 스타일 일부 수정
- 위치 이동 예정

sungik-choi committed:
Feat : 테마 토글 버튼 별도로 분리

- 레이아웃 변경 예정

sungik-choi committed:
Feat : 테마 토글 버튼 변경사항 적용, 아이콘 분리

- 우측 하단 floating action button 스타일로 변경
- theme icon 별도 분리. 일러스트레이터로 제작 후 svg 코드로 적용

sungik-choi committed:
Feat : 토글 버튼 반응형 디자인 추가

- 차후 플로팅 액션 버튼으로 하는 것보다, 모바일 변경시 우측 상단에 버거메뉴가 생기며 네비게이션 하단에 스위치 버튼으로 있는 편이 더 좋을 거 같음
- 플로팅 액션 버튼으로 모바일에서 처리하기엔 너무 중요한 위치를 차지하고 있어서 좋지 않은듯

sungik-choi committed:
Fix : 반응형 디자인 제대로 적용되지 않는 부분 수정

sungik-choi committed:
Style : 다크모드 코드블록 컬러 수정

sungik-choi committed:
Feat : styled-components ThemeContext를 사용해 theme값 전달

- 기존의 ThemeContext 는 삭제

sungik-choi committed:
Refactor : 불필요한 useEffect 제거

- 초깃값을 || 구문으로 설정해도 잘 동작함. 불필요한 useLayoutEffect를 제거

sungik-choi committed:
Fix : 불필요한 조건문 삭제 및 깜빡임 현상 수정

- useLayoutEffect 대신 useEffect 사용으로 깜빡임 현상 사라짐!
- 레이아웃 로딩되기전에 postMessage 메서드 실행되면서 사라지는듯

sungik-choi committed:
Refactor : useEffect 의존성 배열 수정

sungik-choi committed:
Refactor : useEffect 의존성 배열 추가

sungik-choi committed:
Refactor : document -> useRef 로 변경

sungik-choi committed:
Refactor : 함수 매개변수 제거

- 굳이 재사용할 필요가 없는 함수이기에 제거함

sungik-choi committed:
Chore : 패키지 업데이트

sungik-choi committed:
Fix : visually-hidden 스타일 적용 오류 수정

- 미디어 쿼리가 적용되지 않은 코드에 !important가 적용되어서 우선순위가 제대로 적용되지 않는 문제 발생
- !important 코드 추가로 해결

sungik-choi committed:
Feat : ThemeToggleButton 경로 변경 및 메뉴 아이콘 애니메이션 추가

- Background 스타일 별도 분리

sungik-choi committed:
Feat : about 페이지 Layout 컴포넌트 추가

sungik-choi committed:
Feat : useState + transition 활용해 버거 애니메이션 구현

- 애플에서 왜 div를 2개 나눠놨는지 이제 이해했다. 하나는 위치, 하나는 회전값으로 작동해서 애니메이션을 자연스럽게 보이게 만든다.
- useState를 활용해 클릭 시 토글되게 변경하고, 그 값에 따라 transition과 transform 속성이 변경되도록 한다.

sungik-choi committed:
Feat : MenuIcon 컴포넌트 분리

- 불필요한 코드 삭제 후 별도 분리

sungik-choi committed:
Feat : focus 스타일 키보드 접근 시에만 나타나도록 수정

- :focus-visible 수도 클래스 사용

sungik-choi committed:
Refactor : 주석 삭제

sungik-choi committed:
Feat : aria-label 추가 및 button 으로 변경

- 애니메이션 일부 변경

sungik-choi committed:
Feat : 미디어 쿼리 너비를 확인하고 setToggle(false) 만드는 로직 추가

- 화면이 커졌다가 다시 작아지면, 열려있던 메뉴가 닫혀져야하므로 해당되는 로직을 추가했다
- resize + 디바운스로 해결해야하나 싶었지만 matchMedia 와 change 이벤트로 가능하다는 걸 알고 해결했다

sungik-choi committed:
Feat : 메뉴 아이콘 애니메이션 수치, 이름 변경

sungik-choi committed:
Feat : 소형 디바이스 사이즈 추가

sungik-choi committed:
Feat : 메뉴 버튼 클릭 시 내려오는 커튼 컴포넌트 추가

- scaleY 애니메이션으로 동작하는데, 화면이 커졌다 작아졌을 경우 커튼이 아래에서 위로 올라오는 애니메이션이 보이는 현상 발생
- 이를 해결하기 위해 미디어 쿼리가 넘어가는 부분에서 display: none을 일시적으로 줬다가, 500ms 이후 setTimeout에 콜백함수를 넘겨 block으로 변경하도록 처리함

sungik-choi committed:
Feat : 모바일 버거메뉴 메뉴 스타일 추가

sungik-choi committed:
Feat : 테마 토글 버튼 모바일 스타일 적용

- 모바일 사이즈에선 기존 우측 하단 버튼 스타일을 없애고, 다른 링크들과 같은 리스트 스타일로 변경
- 모바일에선 우측 하단 플로팅 버튼으로 계속해서 존재하면 사용성에 영향을 미칠 거라고 판단됐다. 다크<->라이트 모드간 전환이 계속 필요한 게 아니기때문에...

sungik-choi committed:
Fix : 우측에 여백이 생겨 가로스크롤 생기는 문제 해결

sungik-choi committed:
Style : 모바일 메뉴 리스트 폰트 사이즈 축소

sungik-choi committed:
Feat : 테마 토글 버튼 모바일 사이즈에서 아이콘 애니메이션 추가

- 호버 시 아이콘 등장하도록 변경

sungik-choi committed:
Feat : 모바일 메뉴 버튼 a 태그 100% 확장

- 사용성을 위해 클릭 영역을 라인 전체로 확장

sungik-choi committed:
Feat : a 태그 확장 CSS 개선

- 상위 태그에 width 100% 줘서 하위 블록 태그들은 자동으로 100% 차지하도록 변경

sungik-choi committed:
Feat : 모바일 메뉴 리스트들 펼쳐지는 듯한 애니메이션 추가

- 애니메이션 가리기 위해, 커튼 뿐만 아니라 링크 wrap에도 display none -> block 추가

sungik-choi committed:
Feat : 모바일 메뉴 버튼 최상단으로 올림

sungik-choi committed:
Feat : 모바일 메뉴 애니메이션 완료 및 애니메이션 파일 별도 분리

- 애니메이션에 해당하는 코드만 별도로 뽑아 navBarAnimation 파일로 별도 분리

sungik-choi committed:
Style : 테마 토글 버튼 폰트 굵기 변경

sungik-choi committed:
Feat : esc 키로 메뉴 닫는 기능 추가

- useEffect와 keydown 사용. keypress 이벤트는 esc 키를 제대로 입력받지 못했음

sungik-choi committed:
Feat : 네비게이션 바 링크 focus 애니메이션 추가

sungik-choi committed:
Feat : 토글에 따른 pointer-events: none 적용, 링크 ul 높이 재설정

sungik-choi committed:
Feat : 불필요한 코드 삭제, ref 명 변경

sungik-choi committed:
Feat : tabindex를 모바일, 데스크탑 사이즈에 따라 조절하는 로직 추가

- useEffect 2개가 들어갔다. 첫번째는 toggleKeyboardFocus 함수가 생성됐을때, 그리고 toggle이 변경될 때 실행되는 함수다. useCallback의 의존성 배열을 보면 알 수 있다.
toggle이 변경될 때, tabindex를 변경해주는 역할을 맡는다.
- 두번째는 미디어쿼리엘리먼트의 change 이벤트에 바운딩돼있다. 윈도우 사이즈가 변경됐을 때, 데스크탑 사이즈에선 tabindex가 모두 접근돼야되므로 해당 로직을 변경 시 거칠 수 있게 작성해준 부분이다.

sungik-choi committed:
Refactor : 네비게이션 바의 컴포넌트 로직을 useMenu 커스텀 훅으로 분리

sungik-choi committed:
Fix : 하단 여백 생기는 문제 footer를 container 내부로 이동해서 해결

sungik-choi committed:
Fix : 테마 토글 버튼 모바일에서 호버 스타일 유지 수정

- 키보드로 접근시에만 포커싱되게

sungik-choi committed:
Fix : 사파리 크로스 브라우징 문제 해결

- overflow:hidden 때문에 테마 토글버튼 및 커튼 작동안함 해결
- focus-visible 과 hover를 함께 묶으면 사파리에서 호버가 작동하지 않아서 별도로 분리함

sungik-choi committed:
Feat : focus-visible 폴리필 설치 및 적용

- useScript 훅으로 layout에서 적용
- CSS 추가

sungik-choi committed:
Feat : 카테고리 활성화 버튼에 tabindex: -1 추가

- 이미 활성화돼있으므로 키보드로 접근할 필요가 없음. -1을 줘서 접근불가하게 하고 사용자가 혼란스럽지 않게 변경

sungik-choi committed:
Fix : 카테고리 필터 버튼 focus-visible 크로스 브라우징 문제 해결

- 같은 방식으로 중복 코드를 생성해서 해결

sungik-choi committed:
Feat : 미디어쿼리 min-width 에서 max-width 로 변경

- 모바일이 미디어 쿼리 안쪽으로 들어가도록 변경 (데스크탑 우선)
- Content의 width를 100% -> 87.5%로 변경해서 깨지지 않도록 변경

sungik-choi committed:
Fix : 크롬 레이아웃 깨짐 문제 해결

sungik-choi committed:
Feat : 중간 디바이스 사이즈 테마 추가

sungik-choi committed:
Feat : 타이틀 반응형 사이즈 추가

sungik-choi committed:
Feat : 모바일 메뉴 클릭시 trapFocus 기능 추가

- 네비게이션 메뉴 외엔 키보드로 접근이 불가능하도록 설정

sungik-choi committed:
Style : 개행 및 변수명 변경, console.log 삭제

sungik-choi committed:
Fix : 오타 수정

sungik-choi committed:
Feat : html 배경색 추가

- 관성 스크롤 시 하얀색 보이는 현상 해결

sungik-choi committed:
Feat : useScript 삭제 및 관련 코드 삭제

- 사용 안함

sungik-choi committed:
Feat : disable-scroll 패키지 설치

- position: sticky 상태에선 일반적인 body에 overflow:hidden을 줘서 스크롤을 막는 방법이 통하지 않았음.
- 아예 이벤트 자체를 막아버리는 코드로 스크롤 차단

ghost commented:
# [Gatsby Cloud](https://www.gatsbyjs.com) Build Report
**sungik-choi**

:tada: Your build was successful! [See the Deploy preview here.](https://sungikchoi-dev.gtsb.io)

## Build Details

[View the build logs here.](https://gatsbyjs.com/dashboard/0a1f03e2-4319-49d8-9562-417cea95901d/sites/47b582be-70ef-4f85-a003-16aaaf764a14/builds/39273713-25a9-41dc-8e1e-0a2a0fc9c048/details)

:clock1: Build time: **1m**

## Performance

### Lighthouse report

| Metric | Score |
| ----- | ---- |
| Performance |   :green_heart: 100   |
| Accessibility |   :green_heart: 98   |
| Best Practices |   :green_heart: 100   |
| SEO |   :green_heart: 91   |

:link: [View full report](https://build-39273713-25a9-41dc-8e1e-0a2a0fc9c048.gtsb.io/reports/lighthouse/index.html)

sungik-choi committed:
Fix : 갯츠비 빌드 시 window, document 사용하지 못하는 문제 해결 : useTheme

- https://www.gatsbyjs.com/docs/debugging-html-builds/
- 서버 사이드 렌더링 시, window와 document 객체를 초기에 가져올 수 없어 빌드 오류가 발생했다. 이를 해결하기 위해선 오류가 나는 해당 코드들을 useEffect 훅 안으로 옮겨줘야 했다. 코드를 옮겨줘야 해당 객체가 마운트 된 이후에 실행이 되기 때문이다.
- 테마 변경시 깜빡임이 발생하므로 useLayoutEffect 훅을 대신 사용했고, 매치미디어 속성을 useLayoutEffect 훅 안에 넣어주고 초기 1회 테마 셋팅에만 실행되도록 했다.
- 일부분 리팩토링도 함께 진행했다.

sungik-choi committed:
Fix : 갯츠비 빌드 시 window, document 사용하지 못하는 문제 해결 : comment

- 원래 코드에선 useState의 초깃값이 매치미디어 속성으로 바로 설정돼서 문제가 없었는데, 이 코드가 useLayoutEffect로 들어가면서 문제가 발생했다.
- 초기 실행될 땐 theme값이 null, 그리고 이 null값을 바탕으로 themeMode가 만들어지고 comment가 생성돼서 항상 초깃값대로 댓글창이 생성됐다.
- 이를 해결하기위해 useTheme에서 세팅했던 html.dataset을 활용했다. 이쪽은 거의 동기적으로 바로 코드를 가져왔다. theme값이 초기에 null이라 기본값인 light로 설정되는데, dataset을 설정하는 부분은 일련의 과정들을 거치지 않고 빠르게 설정돼서 바로 가져올 수 있는 듯 하다. 이쪽은 더 알아봐야할듯 정확히 어떤 순서로 렌더링이 이루어지는지..
- 초기 1회만 댓글생성을 확실히하고, 초기 테마설정 이후엔 테마값에 따라 테마가 변경되도록 하기 위해 true/false 값을 하나 만들어 댓글 창이 생성될 때 true로 변경되도록 만들었다.

sungik-choi committed:
Fix : 갯츠비 빌드 시 window, document 사용하지 못하는 문제 해결 : useMenu

- window 객체를 사용할 수 없었기 때문에 우선은 useRef(null)에 담아두고, useEffect로 초기에 실행 후 매치미디어를 내부에 담는 식으로 해결했다.

sungik-choi committed:
Feat : 테마 초기값 light 로 변경
